File: server/services/brandMatcher.js
import brands from "../data/legitimateDomains.json" with { type: "json" };
const substitute = value => value.toLowerCase().replace(/0/g, 'o').replace(/1/g, 'l').replace(/3/g, 'e').replace(/4/g, 'a').replace(/5/g, 's').replace(/7/g, 't').replace(/[^a-z0-9]/g, '');
export function levenshtein(left, right) { const row = Array.from({length:right.length+1},(_,i)=>i); for(let i=1;i<=left.length;i++){let prior=row[0];row[0]=i;for(let j=1;j<=right.length;j++){const old=row[j];row[j]=Math.min(row[j]+1,row[j-1]+1,prior+(left[i-1]!==right[j-1]));prior=old;}} return row[right.length]; }
export function findBrandMatch(domain) {
  const candidate = substitute(domain.split('.')[0]); let best = null;
  for (const item of brands) { const token = substitute(item.domain.split('.')[0]); let similarity = 1 - levenshtein(candidate, token) / Math.max(candidate.length, token.length, 1); if (candidate.includes(token) && candidate !== token) similarity = Math.max(similarity, Math.min(.95, .82 + token.length / candidate.length * .1)); if (!best || similarity > best.similarity) best = {...item, similarity: Number(similarity.toFixed(3))}; }
  return best && best.similarity >= .58 ? best : { brand: null, domain: null, similarity: 0 };
}

File: backend/analyzers/brand_matcher.py
import json
from pathlib import Path
from backend.config import REFERENCE_DATA_PATH

SUBSTITUTIONS = str.maketrans({"0": "o", "1": "l", "3": "e", "4": "a", "5": "s", "7": "t"})

def _distance(left: str, right: str) -> int:
    previous = list(range(len(right) + 1))
    for i, a in enumerate(left, 1):
        current = [i]
        for j, b in enumerate(right, 1):
            current.append(min(current[-1] + 1, previous[j] + 1, previous[j - 1] + (a != b)))
        previous = current
    return previous[-1]

def _normalize(value: str) -> str:
    return "".join(c for c in value.lower().translate(SUBSTITUTIONS) if c.isalnum())

def find_brand_match(domain: str, reference_path: Path = REFERENCE_DATA_PATH) -> dict:
    references = json.loads(reference_path.read_text())
    label = domain.split(".")[0]
    candidate = _normalize(label)
    best = {"target_brand": None, "target_domain": None, "brand_similarity": 0.0, "reference_url": None}
    for brand, item in references.items():
        target = _normalize(item["domain"].split(".")[0])
        similarity = 1 - _distance(candidate, target) / max(len(candidate), len(target), 1)
        # A brand token combined with extra login/security wording is more indicative
        # than raw whole-label edit distance alone.
        if target in candidate and candidate != target:
            similarity = max(similarity, min(0.95, 0.82 + (len(target) / max(len(candidate), 1)) * 0.10))
        if similarity > best["brand_similarity"]:
            best = {"target_brand": brand, "target_domain": item["domain"], "brand_similarity": round(similarity, 4), "reference_url": item["url"]}
    return best if best["brand_similarity"] >= 0.58 else {"target_brand": None, "target_domain": None, "brand_similarity": 0.0, "reference_url": None}
